jina/peapods/runtimes/grpc/__init__.py

```diff
@@ -0,0 +1,290 @@
+import argparse
+import asyncio
+import multiprocessing
+import threading
+import time
+from abc import ABC
+from collections import defaultdict
+from typing import Optional, Union, Dict, List
+
+from grpc import RpcError
+
+from jina.enums import OnErrorStrategy
+from jina.excepts import NoExplicitMessage, ChainedPodException, RuntimeTerminated
+from jina.helper import get_or_reuse_loop, random_identity
+from jina.peapods.grpc import Grpclet
+from jina.peapods.runtimes.base import BaseRuntime
+from jina.peapods.runtimes.request_handlers.data_request_handler import (
+    DataRequestHandler,
+)
+from jina.proto import jina_pb2
+from jina.types.message import Message
+from jina.types.routing.table import RoutingTable
+
+
+class GRPCDataRuntime(BaseRuntime, ABC):
+    """Runtime procedure leveraging :class:`Grpclet` for sending DataRequests"""
+
+    def __init__(self, args: argparse.Namespace, **kwargs):
+        """Initialize grpc and data request handling.
+        :param args: args from CLI
+        :param kwargs: extra keyword arguments
+        """
+        super().__init__(args, **kwargs)
+        self._id = random_identity()
+        self._loop = get_or_reuse_loop()
+        self._last_active_time = time.perf_counter()
+
+        self._pending_msgs = defaultdict(list)  # type: Dict[str, List[Message]]
+        self._partial_requests = None
+        self._pending_tasks = []
+
+        self._data_request_handler = DataRequestHandler(args, self.logger)
+        self._grpclet = Grpclet(
+            args=self.args,
+            message_callback=self._callback,
+            logger=self.logger,
+        )
+
+    def _update_pending_tasks(self):
+        self._pending_tasks = [task for task in self._pending_tasks if not task.done()]
+
+    def run_forever(self):
+        """Start the `Grpclet`."""
+        self._grpclet_task = self._loop.create_task(self._grpclet.start())
+        self._loop.run_until_complete(self._grpclet_task)
+
+    def teardown(self):
+        """Close the `Grpclet` and `DataRequestHandler`."""
+        self.logger.debug('Teardown GRPCDataRuntime')
+
+        self._data_request_handler.close()
+        start = time.time()
+        while self._pending_tasks and time.time() - start < 1.0:
```

Comment:
what is the logic about?

Reply:
I am collecting pending async task during the lifetime (they get removed once completed). In the end I want to wait for a second to let them complete. Thats like a simpel implementation of graceful shutdown